### PR TITLE
Added x and xx example, fixes #278. Also fixes tutorial.

### LINKED ISSUE #278: Implement infix:<x> and infix:<xx> in examples/
One as a sub, the other as a macro.

I just went and removed these two operators from core 007 on the master branch. Outside of the tests, the documentation, and the self-hosting, we didn't mention these operators at all. Notably, there was not a single use of them in `examples/`.

But they're highly user-definable, and nothing is lost by defining them this way. In fact, the kind of provide a cute contrast between when you'd use a sub and when you'd use a macro.

(Why is `infix:<xx>` a macro? Because we want the "thunkish" behavior of for example. `my i = 0; my array = i++ xx 3;` (assuming `postfix:<++>` for convenience) should evaluate to `[0, 1, 2]`.)

As #214 points out, it's also quite a good example of an _exception_ to the Single Evaluation Rule &mdash; we get the thunkish behavior precisely because we evaluate the left macro operand once per iteration.

### LINKED ISSUE #287: Expression statements can't begin with {, even in quasi
While trying to debug

```macro swap(a, b) { return quasi { my temp = {{{a}}}; {{{a}}} = {{{b}}}; {{{b}}} = temp; None } }; my w = 5; my z = 10; swap(w, z); say([w, z])```

(the swap example from #194) I found that the error message seems to be due to the statement:expr token's definition being:

```perl6
token statement:expr {
    <![{]>       # }], you're welcome vim
    <EXPR>
}
```

This causes issues in expressions like

```
macro foo (a, b) {return quasi { {{{a}}} + {{{b}}} } };
```

Where the statement expression {{{a}}} + {{{b}}} starts with { and so is not accepted by statement:expr.

This issue also occurs with objects:

```
{a: 2}
```

fails to parse.

I don't feel like I understand perl 6, the parser well enough to know how to fix this issue. I assume that <![{]> is there since we want to accept blocks (which start with {) as blocks rather than expressions, but I'm not sure.

### PR DESCRIPTION
**Pull request by dan-simon**
This PR adds an x and xx example (removing the hello world example). It also updates the tutorial's lists of builtin functions and methods (though it does not include regex or type methods). (The reason I updated the tutorial was that I didn't know about the concat or push list methods and so initially wrote a big mess. I don't want anyone else making the same mistake.)

Potential issues:
* I implemented xx to flatten list input. So 1 xx 3 is [1, 1, 1] and [1, 2] xx 3 is [1, 2, 1, 2, 1, 2]. Is this the desired behavior?
* I don't know how to run all the tests. All I did is create a new test file and run that, while deleting the hello world test file. Are there any other style tests or such I should run?

### CODE REVIEW

**Comment by dan-simon** [comment]
This was meant to be "fixes #278", not "fixes #287", of course.

**Comment by masak** [comment]
I just discovered this PR. Good work! I will comment on lines of code individually, but what I see already looks good enough that I'd like to merge ASAP.

Re your two questions:

* I hesitated in many different directions with respect to flattening. Even had to go back to see how the old, removed `infix:<xx>` worked. (It flattens.) For now we'll let the new one flatten too.
* The proper incantation for running the test suite is `prove -e'perl6 -Ilib' -r t/`. (I'll consider putting it somewhere in the vicinity of the README file.) I have an alias set up locally which runs the `prove` command after first running an empty program in `bin/007` (because this compiles all the module files first before running the test suite, which avoids getting those errors through the test harness).

File: examples/x-and-xx.007 [hunk 1]
```diff
@@ -0,0 +1,39 @@
+macro infix:<xx>(left, right) is equal(infix:<*>) {
+    # Flattens one layer.
+    sub flatten (inputList) {
+        my result = [];
+        for inputList -> i {
+            if (type(i) == type([])) {
```
Comment (masak) [inline examples/x-and-xx.007]: This works, but more idiomatic to use type matching: `if i ~~ Array`.

File: examples/x-and-xx.007 [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+                for i -> j {
+                    result.push(j);
+                }
```
Comment (masak) [inline examples/x-and-xx.007]: Shorter and no less clear than the `for` loop might be to do `result = result.concat(i);` &mdash; this appends all the elements "at once" at the expense of creating a brand new array and letting the old one be collected.

File: examples/x-and-xx.007 [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            } else {
+                result.push(i);
+            }
+        }
+        return result;
+    }
+    return quasi {
+        flatten((^{{{right}}}).map(sub getValue(ignore) {
+            return {{{left}}};
+        }))
```
Comment (masak) [inline examples/x-and-xx.007]: This is beautiful. :heart:

File: examples/x-and-xx.007 [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+# Test number xx number.
+my i = 0;
+say((i = i + 1) xx 3);
+
+# Test list xx number.
+say([1, 2, 3] xx 2);
+
+# Test (list or number) xx number.
+my j = 0;
+say((j = [1, [1, 2, 3]][j && 1]) xx 2);
```
Comment (masak) [inline examples/x-and-xx.007]: No worries for now, but given the way #279 is heading, this expression might become illegal soon. Just a heads-up. :smile: It's an interesting way to check for something &mdash; if we outlaw assignment expressions, we'll have to make sure to have an alternative in place for this one.

**Comment by masak** [comment]
All I made was minor, non-merge-blocking comments, so I'm going to go ahead and merge this now. Again, great work, _especially_ for a first contribution!

**Pull request merged**